model/riscv_insts_zawrs.sail

```diff
@@ -0,0 +1,118 @@
+/*  2. Redistributions in binary form must reproduce the above copyright                 */
+/*     notice, this list of conditions and the following disclaimer in                   */
+/*     the documentation and/or other materials provided with the                        */
+/*     distribution.                                                                     */
+/*                                                                                       */
+/*  THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS''                   */
+/*  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED                    */
+/*  TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A                      */
+/*  PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR                  */
+/*  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,                         */
+/*  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT                     */
+/*  LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF                     */
+/*  USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND                  */
+/*  ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,                   */
+/*  OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT                   */
+/*  OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF                   */
+/*  SUCH DAMAGE.                                                                         */
+/*=======================================================================================*/
+
+/* ****************************************************************** */
+/* This file specifies the instructions in the 'Zawrs' extension.  */
+
+val reservation_set_valid = {
+  ocaml:       "Platform.reservation_set_valid",
+  interpreter: "Platform.reservation_set_valid",
+  c:           "reservation_set_valid",
+  lem:         "reservation_set_valid"
+  } : unit -> bool
+
+/* ****************************************************************** */
+union clause ast = WRS : (wrsop)
+
+mapping clause encdec = WRS(WRS_STO) if haveZawrs()
+  <-> 0b000000011101 @ 0b00000 @ 0b000 @ 0b00000 @ 0b1110011 if haveZawrs()
+
+mapping clause encdec = WRS(WRS_NTO) if haveZawrs()
+  <-> 0b000000001101 @ 0b00000 @ 0b000 @ 0b00000 @ 0b1110011 if haveZawrs()
+
+/* Hart must resume if a locally enabled interrupt becomes pending, even if delegated */
+function interrupts_pending() -> bool = ((mip.bits() & mie.bits()) != zero_extend(0b0))
+
+function clause execute (WRS(op)) = {
+
+  platform_start_wrs_timeout(if op == WRS_STO then true else false);
+
+  while reservation_set_valid() & not(interrupts_pending()) & not(platform_is_wrs_timed_out()) do ();
+
+  platform_stop_wrs_timeout();
+
+  let timed_out : bool = if not(reservation_set_valid()) | interrupts_pending() then false else true;
+
+  match (op, timed_out) {
```

Comment:
You can put this all in one condition. I think this is a bit clearer.

```
if op == WRS_NTO & timed_out & cur_privilege != Machine & mstatus[TW] == 0b1 then {
  handle_illegal(); RETIRE_FAIL
} else RETIRE_SUCCESS
```

Reply:
Its much more explicitly readable documentation of the ISA. This allows one who wants to understand how the ISA is defined to see the conditions clearly spelt out as a truth table  and does not need them to do mental math to evaluate the equation to reconstruct the truth table.

Reply:
I'm fine with the longer version you have but the `let timed_out : bool =` could be:
```
let timed_out : bool = reservation_set_valid() & not(interrupts_pending())
```
i.e. just rewriting `if not(A) | B  then false else true` to `A & not(B)`.

Reply:
updated.

Reply:
> Its much more explicitly readable documentation of the ISA.

IMO my version is more readable and wasn't meant to be "clever". I feel like the `match` version requires careful decoding to understand the actual conditions... But I'm happy to be overruled if people disagree.

You can omit the `: bool` btw. I don't think it adds anything here.

Reply:
I feel the match construct is more explicit as it matches the syntax of a truth table for the various conditions and each row gives the outcome.

Reply:
bool removed.